fedimint-core/src/transaction.rs

```diff
@@ -65,116 +63,37 @@ impl Transaction {
         TransactionId::from_engine(engine)
     }
 
-    /// Validate the aggregated Schnorr Signature signed over the `tx_hash`
-    pub fn validate_signature(
+    /// Validate the schnorr signatures signed over the `tx_hash`
+    pub fn validate_signatures(
         &self,
-        keys: impl Iterator<Item = XOnlyPublicKey>,
+        pub_keys: Vec<XOnlyPublicKey>,
     ) -> Result<(), TransactionError> {
-        let keys = keys.collect::<Vec<_>>();
-
-        // If there are no keys from inputs there are no inputs to protect from
-        // re-binding. This behavior is useful for non-monetary transactions
-        // that just announce something, like LN incoming contract offers.
-        if keys.is_empty() {
-            return Ok(());
+        if pub_keys.len() != self.signatures.len() {
+            return Err(TransactionError::InvalidWitnessLength);
         }
 
-        // Unless keys were empty we require a signature
-        let signature = self
-            .signature
-            .as_ref()
-            .ok_or(TransactionError::MissingSignature)?;
-
-        let agg_pub_key = agg_keys(&keys);
-        let msg =
-            secp256k1_zkp::Message::from_slice(&self.tx_hash()[..]).expect("hash has right length");
-
-        if secp256k1_zkp::global::SECP256K1
-            .verify_schnorr(signature, &msg, &agg_pub_key)
-            .is_ok()
-        {
-            Ok(())
-        } else {
-            Err(TransactionError::InvalidSignature {
-                tx: self.consensus_encode_to_hex().expect("Can't fail"),
-                hash: self.tx_hash().to_hex(),
-                sig: signature.consensus_encode_to_hex().expect("Can't fail"),
-                key: agg_pub_key.consensus_encode_to_hex().expect("Can't fail"),
-            })
+        let txid = self.tx_hash();
+        let msg = secp256k1_zkp::Message::from_slice(&txid[..]).expect("txid has right length");
+
+        for (pk, signature) in pub_keys.iter().zip(&self.signatures) {
+            if secp256k1_zkp::global::SECP256K1
+                .verify_schnorr(signature, &msg, pk)
+                .is_err()
+            {
+                return Err(TransactionError::InvalidSignature {
+                    tx: self.consensus_encode_to_hex().expect("Can't fail"),
+                    hash: self
+                        .tx_hash()
+                        .consensus_encode_to_hex()
+                        .expect("Can't fail"),
+                    sig: signature.consensus_encode_to_hex().expect("Can't fail"),
+                    key: pk.consensus_encode_to_hex().expect("Can't fail"),
```

Comment:
Unrelated to this PR: Lol, why are these `String`s?!

Reply:
Found this confusing as well